src/s_inter.c

```diff
@@ -138,6 +203,514 @@ void sys_set_temppath(void);
+}
+
+// register a callback to delete an object when a send fd fails in the IO
+// thread. This allows to report errors asynchronously
+void sys_addsendfdrmfn(int fd, t_fdsendrmfn fn, void* ptr)
+{
+    int nfd, size;
+    t_fdsend *fs;
+    if (!pd_this->pd_inter->i_fdsend)
+    {
+        /* create an empty FD list */
+        pd_this->pd_inter->i_fdsend = (t_fdsend *)t_getbytes(0);
+        pd_this->pd_inter->i_nfdsend= 0;
+    }
+    if(rbrmfdsend_init())
+        return;
+    nfd = pd_this->pd_inter->i_nfdsend;
+    size = nfd * sizeof(t_fdsend);
+    pd_this->pd_inter->i_fdsend = (t_fdsend *)t_resizebytes(
+        pd_this->pd_inter->i_fdsend, size, size + sizeof(t_fdsend));
+    fs = pd_this->pd_inter->i_fdsend + nfd;
+    fs->fds_fd = fd;
+    fs->fds_fn = fn;
+    fs->fds_ptr = ptr;
+    pd_this->pd_inter->i_nfdsend = nfd + 1;
+}
+
+// a partial replacement for the functionalities of sys_domicrosleep() when
+// enabling THREADED_IO. Namely:
+// - check for ready file descriptors, and for each ready one:
+//  - if it has a RT-friendly callback, retrieve the new data from the
+//  socket and place it in a ring buffer. sys_domicrosleep() will then call the
+//  callback which will retrieve the data from the buffer
+//  - otherwise, mark it so that the callback will be called from
+//  sys_domicrosleep()) at the first occasion
+//
+// As a consequence of this:
+//  - for callbacks that are RT-friendly, sys_domicrosleep() does not need to select()
+//  - for callbacks that are not RT-friendly, sys_domicrosleep() doesn't have to select()
+//  at every call, but only when data is actually available.
+static void poll_fds()
+{
+    const unsigned int bufsize = pd_this->pd_inter->i_iothreadbufsize;
+    ssize_t ret;
+    struct timeval timout = {0}; // making this timeout longer would be more efficient (by avoiding spurious wakeups), but it would needlessly postpone writes
+    int pollem = 1;
+    int i;
+    t_fdpoll *fp;
+    if (pollem && pd_this->pd_inter->i_nfdpoll)
+    {
+        fd_set readset;
+        fdp_select(&readset, timout, kFdpManagerIoThread);
+        for (i = 0; i < pd_this->pd_inter->i_nfdpoll; i++)
+        {
+            t_fdpoll *fp = pd_this->pd_inter->i_fdpoll + i;
+            int fd = fp->fdp_fd;
+            if(fp->fdp_manager != kFdpManagerIoThread)
+                continue;
+            if(!FD_ISSET(fd, &readset))
+                continue;
+            if(fp->fdp_callback_in_audio_thread)
+            {
+                // temporarily transfer ownership
+                SYNC_STORE(&fp->fdp_manager, kFdpManagerAudioThread);
+            }
+            else
+            {
+                int fd = fp->fdp_fd;
+                t_rbskt* rbskt = fp->fdp_rbskt;
+                ring_buffer* rb = rbskt->rs_rb;
+                unsigned int size = rb_available_to_write(rb);
+                if(size > bufsize) {
+                    if(iothreadbuf_resize(size))
+                    {
```

Comment:
I'm not sure I understand the logic here. Why does the buffer have to be greater-equal than the available space in the ringbuffer? On the contrary: buffer is only required to hold a single UDP packet, but the ringbuffer generally has to hold several UDP packets.

Actually, we have to first receive the actual network packet and then check whether the ringbuffer has enough space. The question is what we do if it hasn't. Possible solutions:
1) just send as much as we can and discard the rest. This is obviously not good...
2) somehow manage to flush and grow the ringbuffer (difficult and involves locking)
3) use a second ringbuffer
4) don't call `recv` if the ringbuffer doesn't have space for at least a full UDP packet. The packet will remain in the OS socket buffer.
5) use an unbounded lockfree ringbuffer. This is what I do personally (see https://git.iem.at/cm/aoo/-/blob/develop/common/lockfree.hpp#L207). I would just have to port my C++ code to C, but I guess this is overkill at this stage...

Note that for both 3) and 4), the ringbuffer has to be at least large enough to hold the largest possible UDP packet. Also, the UDP packet will eventually be discarded if Pd can't keep up with reading. The ringbuffer size effectively limits the throughput. This can be seen as either good (can't block audio thread with excessive traffic) or bad (UDP packets are eventually lost if the audio thread can't keep up).

Reply:
> I'm not sure I understand the logic here. Why does the buffer have to be greater-equal than the available space in the ringbuffer?

The reasoning was that I wanted the `rb`'s capacity to be the bottleneck and not, instead, the size of the intermediate storage being used. Clearly there is need for something better.

> Note that for both 3) and 4), the ringbuffer has to be at least large enough to hold the largest possible UDP packet. 

It probably has to be several times that. Currently I am using - fairly arbitrarily - `8 * INBUFSIZE`, because a quick look made me think that if `INBUFSIZE` was good enough for `socketreceiver_read()`, 8x that was good enough as a starting point for this.

> Possible solutions:

What about spinning the thread without `recv()` ing until the audio thread has emptied the ringbuffer? I think this could work for a large-enough ringbuffer.

Reply:
> What about spinning the thread without recv() ing until the audio thread has emptied the ringbuffer? I think this could work for a large-enough ringbuffer.

If you're talking about actual spinning, this is a no-go, as it just burns CPU cycles. But 4) does something very similar, it just doesn't spin. Note that either the ringbuffer has to be large enough to hold the largest possible UDP packet, or you can query the packet size before hand with `int socket_bytes_available(int socket);` (see `s_net.h`)

> 8 * INBUFSIZE

That's 16384 bytes. For most practical situations, this should be fine, but for UDP it could be raised to 65536 to allow huge UDP packets.

Reply:
> If you're talking about actual spinning, this is a no-go, as it just burns CPU cycles. 

Sorry wrong term there, I referred to the thread keep spinning (i.e.: keep calling `sys_doio()` in a loop with sleeps) without actually `recv()`ing. 

> But 4) does something very similar, it just doesn't spin. 

Yes ! 4. is the one, didn't spot it earlier, sorry.

Reply:
Ok, so I would suggest do do 4. This means that for UDP sockets you would us `socket_bytes_available` to check if the ringbuffer has enough space and if not skip the `recv`.

Reply:
And note that for UDP sockets the ringbuffer has to be at least 65536 bytes, otherwise a huge UDP packet will never get through :-)

Reply:
> And note that for UDP sockets the ringbuffer has to be at least 65536 bytes

Or discard/truncate packets larger then the max. ringbuffer capacity.

Reply:
this should be addressed here bf2e29c89321f687944d60603cbfb4ee73a20353

I added an extra level of priority (perhaps conflating things that shouldn't be conflated (i.e.: it is part of the `enum _t_fdp_manager`)), but it should be fine. I don't do atomic read/write there because once we establish that the fdp it is managed by this thread, all the following read/writes are thread safe.

Reply:
@Spacechild1 let me know if you'd prefer this to be in two separate fields of the structure (`atomic_int_t fdp_manager` and `int priority`).

Reply:
Generally, this goes into the right direction. Let's continue here: https://github.com/pure-data/pure-data/pull/1261/commits/b120b03499854e434d96766b701f8b14a6de583a#r561303886